src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs

```diff
@@ -1531,4 +1533,113 @@ private OpenIdConnectProtocolException CreateOpenIdConnectProtocolException(Open
+    {
+        var authHandlerProvider = context.HttpContext.RequestServices.GetRequiredService<IAuthenticationHandlerProvider>();
+        var handler = await authHandlerProvider.GetHandlerAsync(context.HttpContext, context.Scheme.Name);
+        if (handler is OpenIdConnectHandler oidcHandler)
+        {
+            await oidcHandler.HandleValidatePrincipalAsync(context);
+        }
+    }
+
+    private async Task HandleValidatePrincipalAsync(CookieValidatePrincipalContext validateContext)
+    {
+        var accessTokenExpirationText = validateContext.Properties.GetTokenValue("expires_at");
+        if (!DateTimeOffset.TryParse(accessTokenExpirationText, out var accessTokenExpiration))
+        {
+            return;
+        }
+
+        var oidcOptions = this.OptionsMonitor.Get(validateContext.Scheme.Name);
+        var now = oidcOptions.TimeProvider!.GetUtcNow();
+        if (now + TimeSpan.FromMinutes(5) < accessTokenExpiration)
+        {
+            return;
+        }
+
+        var tokenRefreshContext = new Events.TokenRefreshContext(Context, Scheme, oidcOptions, validateContext.Principal!, validateContext.Properties);
+        await Options.Events.TokenRefreshing(tokenRefreshContext);
+        if (tokenRefreshContext.ShouldRefresh)
+        {
+            var oidcConfiguration = await oidcOptions.ConfigurationManager!.GetConfigurationAsync(validateContext.HttpContext.RequestAborted);
+            var tokenEndpoint = oidcConfiguration.TokenEndpoint ?? throw new InvalidOperationException("Cannot refresh cookie. TokenEndpoint missing!");
+
+            using var refreshResponse = await oidcOptions.Backchannel.PostAsync(tokenEndpoint,
+                new FormUrlEncodedContent(new Dictionary<string, string?>()
+                {
+                    ["grant_type"] = "refresh_token",
+                    ["client_id"] = oidcOptions.ClientId,
+                    ["client_secret"] = oidcOptions.ClientSecret,
+                    ["scope"] = string.Join(" ", oidcOptions.Scope),
+                    ["refresh_token"] = validateContext.Properties.GetTokenValue("refresh_token"),
+                }));
+
+            if (!refreshResponse.IsSuccessStatusCode)
+            {
+                validateContext.RejectPrincipal();
+                return;
+            }
+
+            var refreshJson = await refreshResponse.Content.ReadAsStringAsync();
+            var message = new OpenIdConnectMessage(refreshJson);
+
+            var validationParameters = oidcOptions.TokenValidationParameters.Clone();
+            if (oidcOptions.ConfigurationManager is BaseConfigurationManager baseConfigurationManager)
+            {
+                validationParameters.ConfigurationManager = baseConfigurationManager;
+            }
+            else
+            {
+                validationParameters.ValidIssuer = oidcConfiguration.Issuer;
+                validationParameters.IssuerSigningKeys = oidcConfiguration.SigningKeys;
```

Comment:
Does the other parameters need to be added like here?

https://github.com/dotnet/aspnetcore/blob/4141e0e5e2c06b50092c517f089edbf016c261a3/src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs#L1394-L1398